apps/src/applab/AppLabView.jsx

```diff
@@ -34,8 +35,101 @@ var AppLabView = React.createClass({
     onMount: React.PropTypes.func.isRequired
   },
 
+  getInitialState() {
+    // only used so that we can rerender when resized
+    return {
+      width: undefined,
+      height: undefined
+    };
+  },
+
+  /**
+   * Called when the window resizes. Look to see if width/height changed, then
+   * call adjustHeights as our maxHeight may need adjusting.
+   */
+  onResize: function () {
+    var windowWidth = $(window).width();
+    var windowHeight = $(window).height();
+
+    // We fire window resize events when the grippy is dragged so that non-React
+    // controlled components are able to rerender the editor. If width/height
+    // didn't change, we don't need to do anything else here
+    if (windowWidth === this.state.width && windowHeight === this.state.height) {
+      return;
+    }
+
+    this.setState({
+      width: $(window).width(),
+      height: $(window).height()
+    });
+
+    this.adjustHeights();
+  },
+
+  /**
+   * Adjust the height/maxHeight of our instructions based on the rendered size
+   * of the instructions, and the rendered size of the workspace.
+   */
+  adjustHeights: function () {
+    if (!this.props.instructionsInTopPane) {
+      return;
+    }
+
+    var instructionsContent = this.refs.topInstructions.refs.instructions.refs.instructionsMarkdown;
+    var instructionsContentHeight = $(ReactDOM.findDOMNode(instructionsContent)).outerHeight(true);
+    var workspaceHeight = $(ReactDOM.findDOMNode(this.refs.codeWorkspace)).height();
+    var totalHeight = workspaceHeight + this.props.instructionsHeight;
+
+    var height = this.props.instructionsHeight || 300;
+    var maxHeight;
+
+    // Dont want height to extend past rendered length of content.
+    maxHeight = instructionsContentHeight + HEADER_HEIGHT + RESIZER_HEIGHT;
+    if (maxHeight < height) {
+      height = maxHeight;
+    }
+
+    // Have a preference for showing at least 150px of editor and 120px of
+    // debugger. Shrink instructions to make room. If that doesn't provide
+    // enough space, start also shrinking workspace
+    var EDITOR_RESERVE = 150;
+    var DEBUGGER_RESERVE = 120;
+    var INSTRUCTIONS_RESERVE = 150;
```

Comment:
I mean these - pull these and any other constants to the top of this function and use them as part of your 'documentation'... in the hope that someone could adjust them without needing to read the rest of the function.